src/mesh/python/mesh_types.py

```diff
@@ -287,6 +287,314 @@ def __init__(self, bounds_per_dim, num_cells_per_dim):
+                            for tmpridge in ridge_vertices:
+                                if tmpridge[0] == len(vertices)-1 or tmpridge[1] == len(vertices)-1 or \
+                                   tmpridge[0] == ridge[good_vertex] or tmpridge[1] == ridge[good_vertex] or \
+                                   tmpridge[0] == -1 or tmpridge[1] == -1:
+                                    continue
+                                tstpt = ray_segment_intersection(origin, direction,
+                                        [vertices[tmpridge[0]][0], vertices[tmpridge[0]][1]],
+                                        [vertices[tmpridge[1]][0], vertices[tmpridge[1]][1]])
+                                if tstpt is not None:
+                                    bad = True
+                                    break
+                            if not bad:
+                                vertices.append([newpt[0], newpt[1]])
+                                ridge_vertices[n][bad_vertex] = len(vertices)-1
+                                found = True
+                if found == False:
+                    direction[0] = -direction[0]
+                    direction[1] = -direction[1]
+                    for bc in range(4):
+                        pt1 = [bpt1s[bc][0], bpt2s[bc][0]]
+                        pt2 = [bpt1s[bc][1], bpt2s[bc][1]]
+                        for d in range(1,2):
+                            newpt = ray_segment_intersection(origin, direction, pt1, pt2)
+                            if newpt is not None:
+                                # -- check for intersection with other faces
+                                bad = False
+                                for tmpridge in ridge_vertices:
+                                    if tmpridge[0] == len(vertices)-1 or tmpridge[1] == len(vertices)-1 or \
+                                        tmpridge[0] == ridge[good_vertex] or tmpridge[1] == ridge[good_vertex] or \
+                                        tmpridge[0] == -1 or tmpridge[1] == -1:
+                                        continue
+                                    tstpt = ray_segment_intersection(origin, direction,
+                                            [vertices[tmpridge[0]][0], vertices[tmpridge[0]][1]],
+                                            [vertices[tmpridge[1]][0], vertices[tmpridge[1]][1]])
+                                    if tstpt is not None:
+                                        bad = True
+                                        break
+                                if not bad:
+                                    vertices.append([newpt[0], newpt[1]])
+                                    ridge_vertices[n][bad_vertex] = len(vertices)-1
+                                    found = True
+                    if found == False:
+                        print("ERROR No suitable boundary point found!")
+                        sys.exit()
+                continue
```

Comment:
The logic above is quite outrageous but I couldn't think of a simpler way to take a node near the bounding box that needs a node created on the boundary to connect to. A lot can go wrong doing this, e.g. a simple implementation based on the local mesh topology can cast a ray backwards across the entire mesh to intersect with a boundary on the other side of the whole mesh.

Reply:
That sounds familiar...